File: 算法笔记/NQSDemo/src/main/java/jeason/A_Leetcode/其他零散题/_20_有效的括号.java
package jeason.A_Leetcode.其他零散题;
import java.util.HashMap;
import java.util.Stack;

// https://leetcode-cn.com/problems/valid-parentheses/
// 给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串 s ，判断字符串是否有效。
// 有效字符串需满足：
// 左括号必须用相同类型的右括号闭合。
// 左括号必须以正确的顺序闭合。

class _20_有效的括号 {

    private static HashMap<Character, Character> map = new HashMap<>();
    static {
        // key - value
        map.put('(', ')');
        map.put('{', '}');
        map.put('[', ']');
    }
    
    //-------解法1：使用map存取来匹配----------
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        
        int len = s.length();
        for (int i = 0; i < len; i++) {
            char c = s.charAt(i);
            if (map.containsKey(c)) { 
                // 匹配：左括号
                stack.push(c);
            } else { 
                // 匹配：右括号

                //1.没有左括号，无效
                if (stack.isEmpty()) return false;
                
                //2.左右括号匹配不上，无效
                char left = stack.pop();
                if (c != map.get(left)) return false;
            }
        }
        return stack.isEmpty();
    }
    

    //-------解法2：手动判断字符串来匹配----------
    public boolean isValid1(String s) {
        Stack<Character> stack = new Stack<>();
        
        int len = s.length();
        for (int i = 0; i < len; i++) {
            char c = s.charAt(i);
            if (c == '(' || c == '{' || c == '[') { // 左括号
                stack.push(c);
            } else { // 右括号
                if (stack.isEmpty()) return false;
                
                char left = stack.pop();
                if (left == '(' && c != ')') return false;
                if (left == '{' && c != '}') return false;
                if (left == '[' && c != ']') return false;
            }
        }
        return stack.isEmpty();
    }

    //-------解法3：偷懒的解法，while循环删除成对的，最后如果为空就标识是有效的---------
    public boolean isValid2(String s) {
        while (s.contains("{}")
                || s.contains("[]")
                || s.contains("()")) {
            s = s.replace("{}", "");
            s = s.replace("()", "");
            s = s.replace("[]", "");
        }
        return s.isEmpty();
    }
}

File: PythonAlgorithm/Leetcode/栈/_20_有效的括号.py

# https://leetcode.cn/problems/valid-parentheses/

""" 
给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串 s ，判断字符串是否有效。

有效字符串需满足：
1.左括号必须用相同类型的右括号闭合。
2.左括号必须以正确的顺序闭合。
3.每个右括号都有一个对应的相同类型的左括号。

输入：s = "()"
输出：true

输入：s = "()[]{}"
输出：true

输入：s = "(]"
输出：false

题解：https://leetcode.cn/problems/find-all-duplicates-in-an-array/solutions/404530/yuan-di-xiu-gai-yuan-su-wei-fu-shu-zuo-fang-wen-bi/
"""

class Solution(object):

    def isValid(self, s: str) -> bool:
        result: bool = False
        # result = self.isValid1(s)
        # result = self.isValid2(s)
        result = self.isValid3(s)
        return result
    
    # 方法1: 使用字典(哈希表)出入栈判断
    def isValid1(self, s: str) -> bool:
        map: dict = {"{": "}", "[": "]", "(": ")"}
        stack: list = []
        for ch in s:
            if ch in map.keys():
                # 匹配左括号内容(key)
                stack.append(ch)
            else:
                # 匹配右括号内容
                if len(stack) <= 0:
                    return False
                tempItem = stack.pop()
                tempValue: str = map[tempItem]
                if tempValue != ch:
                    return False
        return len(stack) == 0 # 最终一定是栈为空的
    
    # 方法2:手动判断字符串来匹配（太蠢了）
    def isValid2(self, s: str) -> bool:
        stack: list = []
        for ch in s:
            isMatchLeft: bool = (ch == "(") or (ch == "{") or (ch == "[")
            if isMatchLeft:
                # 匹配左括号内容(key)
                stack.append(ch)
            else:
                # 匹配右括号内容
                if len(stack) <= 0:
                    return False
                
                leftItem = stack.pop()
                if (leftItem == "(" and ch != ")") or (leftItem == "[" and ch != "]") or (leftItem == "{" and ch != "}"):
                    return False  
        return len(stack) == 0 # 最终一定是栈为空的
    
    # 方法3:偷懒的解法，while循环删除成对的，最后如果为空就标识是有效的
    def isValid3(self, s: str) -> bool:
        # while条件为非0，就会一直执行
        # find语句找不到时返回-1
        while ((s.find("{}") != -1) or (s.find("[]") != -1) or (s.find("()") != -1)):
            res1 = s.find("{}")
            res2 = s.find("[]")
            res3 = s.find("()")
            print(f"{res1},{res2},{res3}")
            
            s = s.replace("{}", "")
            s = s.replace("[]", "")
            s = s.replace("()", "")
        return len(s) == 0

def run():
    
    tempStr1: str = "{}[]()"
    tempStr2: str = "[]"
    tempStr3: str = "((}}"
    tempStr4: str = "("
    result1: list = Solution().isValid(tempStr1)
    result2: list = Solution().isValid(tempStr2)
    result3: list = Solution().isValid(tempStr3)
    result4: list = Solution().isValid(tempStr4)
    print(f"result1:{result1}\nresult2:{result2}\nresult3:{result3}\nresult4:{result4}")

run()

